### PR TITLE
Support autoFocus

### PR DESCRIPTION
**Pull request by vinhlh**
for #281

### CODE REVIEW

**Comment by vinhlh** [comment]
@n1k0 could you please take a look

**Comment by n1k0** [comment]
@n1k0 I'm off for 3 weeks, I'll take a look when I'm back.

File: src/components/widgets/TextareaWidget.js [hunk 1]
```diff
@@ -1,14 +1,15 @@
 import React, { PropTypes } from "react";
 
 
-function TextWidget({
+function TextareaWidget({
```
Comment (n1k0) [inline src/components/widgets/TextareaWidget.js]: Huh... thanks...

**Comment by n1k0** [comment]
Does this patch work for number fields as well? Selects? We should cover all widgets accepting the autofocus attribute methinks.

**Comment by vinhlh** [comment]
agree, will update.

**Comment by vinhlh** [comment]
@n1k0 sorry, so busy last 2 weeks, could you please help to check again

**Comment by vinhlh** [comment]
what did you mean @sunnyzhang66 
this PR still has no review.

**Comment by sunnyzhang66** [comment]
Hi Vinh,

Thank you for your response.

Currently , I am using CDN of react-jsonschema-form which is

<script src="https://unpkg.com/react-jsonschema-form/dist/react-jsonschema-form.js"></script>

, but doesn't have auto focus code inside.

I wonder if you can upgrade CDN which is including latest code?



Thanks

Sunny


________________________________

**Comment by vinhlh** [comment]
@sunnyzhang66 this commit is still in an `open` Pull Request.

File: src/components/widgets/BaseInput.js [hunk 2]
```diff
@@ -18,6 +19,7 @@ function BaseInput(props) {
       {...inputProps}
       className="form-control"
       readOnly={readonly}
+      autoFocus={autoFocus}
```
Comment (n1k0) [inline src/components/widgets/BaseInput.js]: I think we went with the convention of lowercasing html attribute names, as you can see with `readOnly -> readonly` above. Could you please rename `autoFocus` to `autofocus`?

File: test/uiSchema_test.js [hunk 3]
```diff
@@ -247,6 +247,112 @@ describe("uiSchema", () => {
...
+      it("should focus on multiple files input", () => {
+        shouldFocus({type: "array", items: {type: "string", format: "data-url"}}, {"ui:autoFocus": true});
+      });
+
+      it("should focus on first item of a list of strings", () => {
+        shouldFocus({type: "array", items: {type: "string", default: "foo"}}, {"ui:autoFocus": true}, "input", ["foo", "bar"]);
+      });
+
+      it("should focus on first item of a multiple choices list", () => {
+        shouldFocus({type: "array", items: {type: "string", enum: ["foo", "bar"]}, uniqueItems: true}, {"ui:widget": "checkboxes", "ui:autoFocus": true}, "input", ["bar"]);
```
Comment (n1k0) [inline test/uiSchema_test.js]: We usually try to not exceed 120c. line length in this project (maybe we should enforce this rule in eslint config)

File: README.md [hunk 4]
```diff
@@ -521,6 +522,18 @@ const uiSchema = {
 
 Help texts work for any kind of field at any level, and will always be rendered immediately below the field component widget(s), but after contextualized errors, if any.
 
+### Auto focus
+
+If you want to focus on a text input or textarea input automatically, just set `ui:autoFocus` uiSchema directive to `true`.
```
Comment (n1k0) [inline README.md]: s/on a text input or textarea input/on a widget

File: src/components/widgets/BaseInput.js (hunk 2 above)
Comment (vinhlh) [inline src/components/widgets/BaseInput.js]: @n1k0 I am curious about the reason why we need lowercasing html atrribute names, could you help to explain

**Comment by vinhlh** [comment]
just push a fix, @n1k0 could you help to take a look

File: src/components/widgets/BaseInput.js (hunk 2 above)
Comment (vinhlh) [inline src/components/widgets/BaseInput.js]: might be: to distinguish with normal props. I have updated.

File: README.md [hunk 5]
```diff
@@ -524,7 +524,7 @@ Help texts work for any kind of field at any level, and will always be rendered
 
 ### Auto focus
 
-If you want to focus on a text input or textarea input automatically, just set `ui:autoFocus` uiSchema directive to `true`.
+If you want to focus on a text input or textarea input/on a widget automatically, just set `ui:autoFocus` uiSchema directive to `true`.
```
Comment (n1k0) [inline README.md]: Nit: `autofocus` now we've updated the directive name

File: README.md [hunk 6]
```diff
@@ -521,6 +522,18 @@ const uiSchema = {
 
 Help texts work for any kind of field at any level, and will always be rendered immediately below the field component widget(s), but after contextualized errors, if any.
 
+### Auto focus
+
+If you want to focus on a text input or textarea input/on a widget automatically, just set `ui:autoFocus` uiSchema directive to `true`.
```
Comment (n1k0) [inline README.md]: `ui:autofocus` now we've updated the directive name case

File: README.md [hunk 7]
```diff
@@ -521,6 +522,18 @@ const uiSchema = {
...
+
+```js
+const schema = {type: "string"};
+const uiSchema = {
+  "ui:widget": "textarea",
+  "ui:autoFocus": true
```
Comment (n1k0) [inline README.md]: `ui:autofocus`

**Comment by vinhlh** [comment]
@n1k0 sorry, i forgot. Updated.

**Comment by n1k0** [comment]
Thanks a lot for your contribution!

**Comment by n1k0** [comment]
Released in [v0.41.0](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.41.0).

**Pull request merged**